src/funPieceListLog.cpp

```diff
@@ -1206,3 +1206,840 @@ void PiecewisePoissonLossLog::push_piece
+                 first_mean,
+                 last_min_mean, first_max_mean);
+        }
+      }
+    }
+  }//if(two_roots
+  if(second_mean != INFINITY){
+    // two crossing points.
+    double before_mean = (last_min_mean + first_mean )/2;
+    double cost_diff_before = diff_piece.getCost(before_mean);
+    if(cost_diff_before < 0){
+      push_piece(it1, last_min_mean, first_mean);
+      push_piece(it2, first_mean, second_mean);
+      push_piece(it1, second_mean, first_max_mean);
+    }else{
+      push_piece(it2, last_min_mean, first_mean);
+      push_piece(it1, first_mean, second_mean);
+      push_piece(it2, second_mean, first_max_mean);
+    }
+    if(verbose)printf("not equal on the sides, 2 crossing points\n");
+  }else if(first_mean != INFINITY){
+    // "one" crossing point. actually sometimes we have last_min_log_mean
+    // < first_log_mean < first_max_log_mean but cost_diff_before and
+    // cost_diff_after have the same sign! In that case we need to
+    // just push one piece.
+    double before_mean = (last_min_mean + first_mean) / 2;
+    double cost_diff_before = diff_piece.getCost(before_mean);
+    if(verbose){
+      printf("cost_diff_before(%.55f)=%f\n", before_mean, cost_diff_before);
+    }
+    double after_mean = (first_max_mean + first_mean)/2;
+    double cost_diff_after = diff_piece.getCost(after_mean);
+    if(verbose)printf("cost_diff_after(%.55f)=%f\n", after_mean, cost_diff_after);
+    if(cost_diff_before < 0){
+      if(cost_diff_after < 0){
+        // f1-f2<0 meaning f1<f2 on the entire interval, so just push it1.
+        push_piece(it1, last_min_mean, first_max_mean);
+      }else{
+        push_piece(it1, last_min_mean, first_mean);
+        push_piece(it2, first_mean, first_max_mean);
+      }
+    }else{//f1(before)-f2(before)>=0 meaning f1(before)>=f2(before)
+      if(cost_diff_after < 0){
+        //f1(after)-f2(after)<0 meaning f1(after)<f2(after)
+        push_piece(it2, last_min_mean, first_mean);
+        push_piece(it1, first_mean, first_max_mean);
+      }else{
+        //f1(after)-f2(after)>=0 meaning f1(after)>=f2(after)
+        push_piece(it2, last_min_mean, first_max_mean);
+      }
+    }
+    if(verbose)printf("not equal on the sides, 1 crossing point\n");
+  }else{
+    // "zero" crossing points. actually there may be a crossing point
+    // in the interval that is numerically so close as to be identical
+    // with last_min_log_mean or first_max_log_mean.
+    if(verbose){
+      printf("not equal on the sides, zero crossing points\n");
+      printf("cost_diff left=%e mid=%e right=%e\n",
+             cost_diff_left, cost_diff_mid, cost_diff_right);
+    }
+    double cost_diff;
+    if(first_max_mean == INFINITY){
+      // if we are at the last interval and the right limit is
+      // infinity, then it should be fine to compare the cost at any
+      // point in the interval.
+      cost_diff = diff_piece.getCost(last_min_mean + 1); // WHY + 1
```

Comment:
Thanks, Toby! It is not completely clear to me that it was inside an explicit conditional, but it may have been handled implicitly earlier in your code that pushes a new piece. At any rate, your comment prompted me to find a few others things I should be checking! I'll let you know if I have any other questions. Thanks again!